stdlib/public/Concurrency/Task.swift

```diff
@@ -0,0 +1,273 @@
+  ///   - this DOES NOT affect `Task.currentPriority()`.
+  /// - if a task is created with a `Task.Handle`, and a higher-priority task
+  ///   calls the `await handle.get()` function the priority of this task must be
+  ///   permanently increased until the task completes.
+  ///   - this DOES affect `Task.currentPriority()`.
+  ///
+  /// TODO: Define the details of task priority; It is likely to be a concept
+  ///       similar to Darwin Dispatch's QoS; bearing in mind that priority is not as
+  ///       much of a thing on other platforms (i.e. server side Linux systems).
+  public struct Priority: Comparable {
+    public static let `default`: Task.Priority = .init() // TODO: replace with actual values
+
+    // TODO: specifics of implementation are not decided yet
+    private let __value: Int = 0
+
+    public static func < (lhs: Self, rhs: Self) -> Bool {
+      lhs.__value < rhs.__value
+    }
+  }
+}
+
+// ==== Task Handle ------------------------------------------------------------
+
+extension Task {
+
+  /// A task handle refers to an in-flight `Task`,
+  /// allowing for potentially awaiting for its result or canceling it.
+  ///
+  /// It is not a programming error to drop a handle without awaiting or canceling it,
+  /// i.e. the task will run regardless of the handle still being present or not.
+  /// Dropping a handle however means losing the ability to await on the task's result
+  /// and losing the ability to cancel it.
+  public final class Handle<Success, Failure: Error> {
+      /// Wait for the task to complete, returning (or throwing) its result.
+      ///
+      /// ### Priority
+      /// If the task has not completed yet, its priority will be elevated to the
+      /// priority of the current task. Note that this may not be as effective as
+      /// creating the task with the "right" priority to in the first place.
+      public func get() async throws -> Success {
+      fatalError("\(#function) not implemented yet.")
+    }
+
+    /// Attempt to cancel the task.
+    ///
+    /// Whether this function has any effect is task-dependent.
+    ///
+    /// For a task to respect cancellation it must cooperatively check for it
+    /// while running. Many tasks will check for cancellation before beginning
+    /// their "actual work", however this is not a requirement nor is it guaranteed
+    /// how and when tasks check for cancellation in general.
+    public func cancel() {
+      fatalError("\(#function) not implemented yet.")
+    }
+  }
+}
+
+extension Task.Handle where Failure == Never {
+  /// Wait for the task to complete, returning its result.
+  ///
+  /// ### Priority
+  /// If the task has not completed yet, its priority will be elevated to the
+  /// priority of the current task. Note that this may not be as effective as
+  /// creating the task with the "right" priority to in the first place.
+  public func get() async -> Success {
```

Comment:
Should this go away?

Reply:
Ah damn, forgot cancelation!

Thanks Doug, will remove

Reply:
Actually... that's not clear yet!

The more I dug into this while writing up the proposal the more second thoughts I was having.

Pasing example semantics comparison for reference:


```swift
  let neverChecks: Task.Handle<String, Never> = Task.runDetached {
    // never checks cancellation
    // Task.sleep(.seconds(60)) // pretend being slow
    "ok"
  }
​
  // ----
​
  // Semantics A) Cancellation truly is only co-operative
  //              These semantics would argue FOR the existence of the non-throwing get()
  neverChecks.cancel() // we cancel immediately
  // get ALWAYS throws, i.e. we're expecting this to throw a CancellationError
  // but under Semantics A won't -- we're truly cooperative, and the task never
  // checked cancellation so in reality it is as if it wasn't cancelled.
  await try neverChecks.get()
  // because the task never checked cancellation, we get "ok here"
  // Semantics A conclusion: we should offer a non throwing get()
​
  // ----
​
  // Semantics B) Cancellation is eager on handles, hard and does NOT guarantee emitting the same error
  //              These semantics would argue AGAINST the existence of the non throwing get()
  neverChecks.cancel() // we cancel immediately
  await try neverChecks.get() // under "eager" B cancellation semantics this will immediately throw
  // even if the task still executed and completed successfully (because it never checked for cancellation!)
  // this would throw a CancellationError under Semantics B
```

I would argue semantics A is what people are meaningfully after, and it allows building more predictably on the tasks / handles. This would imply that we _do_ need the non throwing get.